File: javascript/json-transform/src/__tests__/functions/TransformerFunctionMin.test.ts
import type BigNumber from "bignumber.js";
import { describe, test } from "vitest";
import { assertTransformation } from "../BaseTransformationTest";
import { BigDecimal } from "../../functions/common/FunctionHelpers";

describe("TransformerFunctionMin", () => {
  test("inline", async () => {
    const arr = [4, -2, 13.45, null];
    await assertTransformation(arr, "$$min($$long:-4):$", BigInt(-4));
    await assertTransformation(arr, "$$min(-8,NUMBER):$", "-8");
    await assertTransformation(arr, "$$min():$", null);
    await assertTransformation(arr, "$$min(z,STRING):$", -2);
  });

  class Holder {
    public value: BigNumber | null;
    constructor(value: BigNumber | null) {
      this.value = value;
    }
  }

  test("object", async () => {
    const arr = [
      new Holder(BigDecimal(4)),
      new Holder(BigDecimal(2)),
      new Holder(BigDecimal("13.45")),
      new Holder(null),
    ];
    await assertTransformation(
      arr,
      {
        $$min: "$",
        by: "##current.value",
      },
      null,
    );

    await assertTransformation(
      arr,
      {
        $$min: "$",
        by: "##current.value",
        default: 1,
      },
      1,
    );
  });
});

File: java/json-transform/src/test/java/co/nlighten/jsontransform/functions/TransformerFunctionMinTest.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.BaseTest;
import com.google.gson.JsonNull;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;

public class TransformerFunctionMinTest extends BaseTest {
    @Test
    void inline() {
        var arr = new Object[] {4,-2,13.45, JsonNull.INSTANCE};
        assertTransformation(arr, "$$min($$long:-4):$", fromJson("-4"));
        assertTransformation(arr, "$$min(-8,NUMBER):$", fromJson("-8"));
        assertTransformation(arr, "$$min():$", fromJson(null));
        assertTransformation(arr, "$$min(z,STRING):$", fromJson("-2"));

    }

    record holder(BigDecimal value) {}
    @Test
    void object() {
        var arr = new holder[] {new holder(new BigDecimal(4)),new holder( new BigDecimal(2)),new holder( new BigDecimal("13.45")), new holder(null)};


        assertTransformation(arr, fromJson("""
{
  "$$min":"$","by":"##current.value"
}
"""), fromJson("null"));

        assertTransformation(arr, fromJson("""
{
  "$$min":"$","by":"##current.value","default":1
}
"""), fromJson("1"));

    }

}
